[ConstitutiveLawsApp] High cycle fatigue law update

SergioJimenezReyes opened the pull request:
**📝 Description**
This PR updates the High Cycle Fatigue (HCF) Constitutive Law (CL) and related components to improve flexibility, robustness, and consistency with current code practices. It introduces enhanced behavior handling for complex load cases, integrates recent research-based improvements, and aligns with the structure used in other CLs.

**🆕 Changelog**
- Updated the HCF CL to follow the current style and structure (based on the isotropic damage CL)
- Added support to distinguish monotonic and cyclic load behaviors in the CL
- Introduced handling for reference damage states with potential multiple large jumps in load history
- Implemented smoothness adjustment of the fatigue reduction parameter based on CFactor (following S. Jiménez's thesis: [link](https://upcommons.upc.edu/handle/2117/421446))
- Updated the HCF container to be compatible with the new CL structure
- Extended AdvanceInTimeProcess to support the new fatigue law behavior
- Updated related cpp and Python tests accordingly

_Note: This PR is the first of a series. Upcoming smaller PRs will address:_
- Generalization of the law for 2D/3D cases
- Initial stress/strain feature
- Minor fixes and improvements in AdvanceInTimeProcess
- Further enhancements to the HCF container

SergioJimenezReyes committed:
CL update according to damage CL

SergioJimenezReyes committed:
Update of the HCF CL

SergioJimenezReyes committed:
Advance in time strategy update

Now it is prepared to account for multiple nonlinearities taking place simulataneously (damage and plasticity). It also distinguishes between two types of loads: monotonic and cyclic

SergioJimenezReyes committed:
Minimum update to HCF container

Compatibilizing the container with the new features in HCF CL. A final update is needed then to reach the same state and substitute most features in fatigue CL. This will be tackled in the future.

SergioJimenezReyes committed:
Merge branch 'master' into HCF_UpdatePhD

SergioJimenezReyes committed:
Updating hcf analysis stage

It is necessary this update to make this compatible with the (in)active AIT in a fatigue case and detect if the applied load is monotonic or cyclic

SergioJimenezReyes committed:
Updating cpp tests

SergioJimenezReyes committed:
Cleaning

SergioJimenezReyes committed:
Update HCF python test

Updating the test now including check for AIT process (first jump and jump during damage phase). Tests are update to allow different analysis stages

SergioJimenezReyes committed:
Merge branch 'master' into HCF_UpdatePhD

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_constitutive/auxiliary_files/cl_integrators/high_cycle_fatigue_law_integrator.h:
```diff
@@ -204,22 +193,51 @@ class HighCycleFatigueLawIntegrator
+        const Vector& r_fatigue_coefficients = rMaterialParameters[HIGH_CYCLE_FATIGUE_COEFFICIENTS];
+        double number_of_cycles_to_failure = Nf;
+        const double BETAF = r_fatigue_coefficients[4];
+        if (UltimateStress - MaxStress > 1.0e-4) {
```
we can unify this tol as a `constexpr double`, FYI

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_constitutive/small_strains/fatigue/generic_small_strain_high_cycle_fatigue_law.cpp:
```diff
@@ -96,21 +97,57 @@ void GenericSmallStrainHighCycleFatigueLaw<TConstLawIntegratorType>::InitializeM
     double cycles_to_failure = mCyclesToFailure;
-    bool adnvance_strategy_applied = rValues.GetProcessInfo()[ADVANCE_STRATEGY_APPLIED];
-    bool damage_activation = rValues.GetProcessInfo()[DAMAGE_ACTIVATION];
+    bool advance_in_time_process_applied = rValues.GetProcessInfo()[ADVANCE_STRATEGY_APPLIED];
```
we can create a `r_process_info = rValues.GetProcessInfo()` and reuse it

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_constitutive/small_strains/fatigue/generic_small_strain_high_cycle_fatigue_law.cpp:
```diff
@@ -96,21 +97,57 @@ void GenericSmallStrainHighCycleFatigueLaw<TConstLawIntegratorType>::InitializeM
+
+    if (max_indicator && min_indicator && current_load_type) {
+        if (mFirstCycleOfANewLoad) {
+            const SizeType fatigue_parameters_size = rValues.GetMaterialProperties()[HIGH_CYCLE_FATIGUE_COEFFICIENTS].size();
```
we can `const auto& r_mat_props = rValues.GetMaterialProperties()` and reuse

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_processes/advance_in_time_high_cycle_fatigue_process.cpp:
```diff
@@ -84,6 +86,87 @@ void AdvanceInTimeHighCycleFatigueProcess::Execute()
+    auto& process_info = mrModelPart.GetProcessInfo();
+    double time = process_info[TIME];
+    double delta_time;
+    if (mThisParameters["solver_settings"]["time_stepping"].Has("time_step")) {
```
This is wrong IMO. We should take the current time step from the process info as `r_process_info[DELTA_TIME]`

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_processes/advance_in_time_high_cycle_fatigue_process.cpp:
```diff
@@ -84,6 +86,87 @@ void AdvanceInTimeHighCycleFatigueProcess::Execute()
+                if (monotonic_constraints_list[i] == model_part_name && time >= model_part_start_time && time <= model_part_end_time && !break_condition) {
+                    break_condition = true;
+                    //Checking if this is the first step of a new model part
+                    double model_part_start_time = mThisParameters["processes"]["constraints_process_list"][j]["Parameters"]["interval"][0].GetDouble();
```
we just declared this variable a few lines above, is this ok?

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_processes/advance_in_time_high_cycle_fatigue_process.cpp:
```diff
@@ -84,6 +86,87 @@ void AdvanceInTimeHighCycleFatigueProcess::Execute()
+        std::vector<std::string> monotonic_constraints_list = mThisParameters["fatigue"]["monotonic_constraints_process_list"].GetStringArray();
+        //Loop on the monotonic constraints list
+        for (unsigned int i = 0; i < monotonic_constraints_list.size(); i++) {
+            for (unsigned int j = 0; j < mThisParameters["processes"]["constraints_process_list"].size(); j++) {
```
instead of `unsigned int` now we commonly use `IndexType`

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_processes/advance_in_time_high_cycle_fatigue_process.cpp:
```diff
@@ -84,6 +86,87 @@ void AdvanceInTimeHighCycleFatigueProcess::Execute()
+    if (has_cyclic_constraints_list) {
+        std::vector<std::string> cyclic_constraints_list = mThisParameters["fatigue"]["cyclic_constraints_process_list"].GetStringArray();
+        //Loop on the cyclic constraints list
+        for (unsigned int i = 0; i < cyclic_constraints_list.size(); i++) {
```
IDEM

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_processes/advance_in_time_high_cycle_fatigue_process.cpp:
```diff
@@ -123,9 +206,75 @@ void AdvanceInTimeHighCycleFatigueProcess::CyclePeriodPerIntegrationPoint(bool&
+    std::vector<double> previous_cycle_damage;
+    std::vector<double> plastic_dissipation;
+    std::vector<double> previous_cycle_plastic_dissipation;
+    std::vector<double> plastic_damage;
```
damage and dissipation are ad hoc internal variables. Maybe we could think how to name them in a general form. TODO in the future

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_processes/advance_in_time_high_cycle_fatigue_process.cpp:
```diff
@@ -123,9 +206,75 @@ void AdvanceInTimeHighCycleFatigueProcess::CyclePeriodPerIntegrationPoint(bool&
+            if (is_damage) {
+                r_elem.CalculateOnIntegrationPoints(DAMAGE, damage, process_info);
+                r_elem.CalculateOnIntegrationPoints(PREVIOUS_CYCLE_DAMAGE, previous_cycle_damage, process_info);
+                for (unsigned int i = 0; i < number_of_ip; i++) {
```
IDEM

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_processes/advance_in_time_high_cycle_fatigue_process.cpp:
```diff
@@ -155,35 +304,32 @@ void AdvanceInTimeHighCycleFatigueProcess::StableConditionForAdvancingStrategy(b
         }
     }
-    if ((acumulated_max_stress_rel_error < 1e-4 && acumulated_rev_factor_rel_error < 1e-4 && fatigue_in_course) || (DamageIndicator && acumulated_max_stress_rel_error < 1e-3 && acumulated_rev_factor_rel_error < 1e-3 && fatigue_in_course)) {
+    if ((acumulated_max_stress_rel_error < 1e-4 && acumulated_rev_factor_rel_error < 1e-4 && fatigue_in_course) || (NoLinearityIndicator && acumulated_max_stress_rel_error < 1e-3 && acumulated_rev_factor_rel_error < 1e-2 && fatigue_in_course)) {
```
TODO: make these indicators relative to the amount of IPs
TODO: expose the tolerance to the json

SergioJimenezReyes committed:
Typedef BoundedArrayType and tolerance

Addressing PR

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_constitutive/auxiliary_files/cl_integrators/high_cycle_fatigue_law_integrator.h:
```diff
@@ -115,10 +118,10 @@ class HighCycleFatigueLawIntegrator
         const double stress_increment_1 = stress_1 - stress_2;
         const double stress_increment_2 = CurrentStress - stress_1;
-        if (stress_increment_1 > 1.0e-3 && stress_increment_2 < -1.0e-3) {
+        if (stress_increment_1 > tolerance && stress_increment_2 < -tolerance) {
```
careful because now you've decreased this tolerance from `1e-3` to the minimum tolerance of the computer...

SergioJimenezReyes committed:
Grouping mat_props and process_info

SergioJimenezReyes committed:
Eliminating duplicated code

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_constitutive/auxiliary_files/cl_integrators/high_cycle_fatigue_law_integrator.h:
```diff
@@ -205,7 +208,8 @@ class HighCycleFatigueLawIntegrator
             rB0 = -(std::log(MaxStress / UltimateStress) / std::pow((std::log10(rNf)), FatigueReductionFactorSmoothness * square_betaf));
         } else {
-            rNf = 1.0e20; // No fatigue at this IP, i.e., proposing infinite jump.
+            rNf = std::numeric_limits<double>::infinity(); // No fatigue at this IP, i.e., proposing infinite jump.
```
nice

SergioJimenezReyes committed:
unsigned int to IndexType

SergioJimenezReyes committed:
Sorting DELTA_TIME issue and improving code

I am allowing to use as input both "End" and a double for the end time in constraint_process_list

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_processes/advance_in_time_high_cycle_fatigue_process.cpp:
```diff
@@ -137,16 +149,34 @@ void AdvanceInTimeHighCycleFatigueProcess::MonotonicOrCyclicLoad()
                     //Checking if this is the first step of a new model part
                     if (time - delta_time <= model_part_start_time) {
                         new_model_part = true;
+                        KRATOS_WATCH("SIII")
```
idem

AlejandroCornejo commented on applications/ConstitutiveLawsApplication/custom_processes/advance_in_time_high_cycle_fatigue_process.cpp:
```diff
@@ -111,19 +103,39 @@ void AdvanceInTimeHighCycleFatigueProcess::MonotonicOrCyclicLoad()
                     //Checking if this is the first step of a new model part
                     if (time - delta_time <= model_part_start_time) {
                         new_model_part = true;
+                        KRATOS_WATCH("YESS")
```
NOO ;)

SergioJimenezReyes committed:
Removing KRATOS_WATCH

SergioJimenezReyes committed:
Merge branch 'master' into HCF_UpdatePhD
